### PR TITLE
CDRIVER-4679 Prefer XSI-compliant strerror_r when available

### PR DESCRIPTION
**Pull request by eramongodb**
Attempts to resolve CDRIVER-4679 by preferring the XSI-compliant version of `strerror_r` when available by using a more accurate feature test macros condition. The GNU extension is used as a fallback instead. An additional `#error` was added to hopefully improve the resulting compilation error if a supported `strerror_r` implementation cannot be found.

Verified by [this patch](https://spruce.mongodb.com/version/64b8210ac9ec44c255145475). Note: _not_ yet verified on Alpine Linux as described in CDRIVER-4679 as the corresponding EVG distro is not available, so not entirely confident this actually addresses the originally reported issue.

### CODE REVIEW

File: src/libbson/src/bson/bson-error.c [hunk 1]
```diff
@@ -109,12 +109,17 @@ bson_strerror_r (int err_code,  /* IN */
    if (strerror_s (buf, buflen, err_code) != 0) {
       ret = buf;
    }
-#elif defined(__GNUC__) && defined(_GNU_SOURCE)
-   ret = strerror_r (err_code, buf, buflen);
-#else /* XSI strerror_r */
+#elif defined(_XOPEN_SOURCE) && _XOPEN_SOURCE >= 600 && !defined(_GNU_SOURCE)
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libbson/src/bson/bson-error.c:112]: > What is the argument for not using the logic suggested in [strerror_r(3)](https://linux.die.net/man/3/strerror_r)?

`_XOPEN_SOURCE` implies `_POSIX_C_SOURCE` per https://pubs.opengroup.org/onlinepubs/7908799/xsh/compilation.html: "Since this specification is aligned with the ISO C standard, and since all functionality enabled by _POSIX_C_SOURCE set greater than zero and less than or equal to 199506L should be enabled by _XOPEN_SOURCE set equal to 500, there should be no need to define either _POSIX_SOURCE or _POSIX_C_SOURCE if _XOPEN_SOURCE is so defined."

> Likewise, `!_GNU_SOURCE` would also handle the case where `_GNU_SOURCE` is defined but zero (not sure how relevant that is).

Only the presence of `_GNU_SOURCE` matters, not its value, per https://man7.org/linux/man-pages/man7/feature_test_macros.7.html: "Defining this macro (with any value) implicitly defines [...]. In addition, various GNU-specific extensions are also exposed."

> I don't see `_POSIX_SOURCE` referred to anywhere within libmongoc.

It is unconditionally defined via the definition of `_XOPEN_SOURCE` here: https://github.com/mongodb/mongo-c-driver/blob/6d2be25206daeb55960c84a5f129b226f0a25393/CMakeLists.txt#L398

All our requirements concerning POSIX and extensions are being handled by the (unconditional) definition of the macros listed in the link above. The conflict between the XSI-compliant version and GNU extension version of `strerror_r` is AFAIK a unique case.

[the remark this answers is not in the document]
Comment (jmikola) [reply src/libbson/src/bson/bson-error.c:112]: Thanks for the rigorous response! LGTM.

**Review by jmikola — APPROVED** [review]

**Review by kevinAlbs — COMMENTED** [review]
musl [defines the XSI-conforming strerror_r](https://git.musl-libc.org/cgit/musl/tree/include/string.h#n66) when `_GNU_SOURCE` is defined.

Compiling with Alpine with docker results in an error:
```
docker run --rm -it --platform linux/amd64 -v `pwd`:/c-driver alpine
```

Configured with `_GNU_SOURCE` defined:

```
cmake \
    -DCMAKE_C_FLAGS="-D_GNU_SOURCE=1" \
    -DBUILD_VERSION="1.25.0-pre" \
    -DENABLE_SSL=OFF \
    -DENABLE_SRV=OFF \
    -S$(pwd) -B$(pwd)/cmake-build
```

Building selects the GNU extension:

```
/c-driver/src/libbson/src/bson/bson-error.c: In function 'bson_strerror_r':
/c-driver/src/libbson/src/bson/bson-error.c:120:8: error: assignment to 'char *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  120 |    ret = strerror_r (err_code, buf, buflen);
      |        ^
[ 40%] Building C object src/libbson/CMakeFiles/bson_shared.dir/src/bson/bson-iter.c.o
```

Does this suggest a bug in musl? Is musl be expected to provide the GNU extension when `_GNU_SOURCE` is defined?

**Comment by eramongodb** [comment]
> Does this suggest a bug in musl? Is musl be expected to provide the GNU extension when _GNU_SOURCE is defined?

This seems to be a deliberate choice by musl, not a bug. Support for `_GNU_SOURCE` (in a meaningful way) seems to be a hard "no", per [this mailing thread](https://www.openwall.com/lists/musl/2018/08/05/2):

> > I have found that _GNU_SOURCE is defined when I build snort on OpenWrt (OpenWrt uses musl). This causes the build to fail because snort expects strerror_r to return a char * since _GNU_SOURCE is defined.
> > 
> > Is this a bug in musl?
>
> No, musl explicitly does not [support] the GNU interfaces that conflict with standard interfaces by the same name.

This issue seems to have also plagued Gentoo maintainers as documented [here](https://blogs.gentoo.org/gsoc2016-native-clang/2016/06/15/the-invisible-_gnu_source-in-your-c-code/):

> Should we blame musl for this? [...] Unfortunately the ugliness doesn’t just end here. It turns out g++/clang++ unconditionally predefines `_GNU_SOURCE` when compiling any C++ code on Linux, and that’s because libc++/libstd++ won’t work on Linux without this macro defined. So the perfect fix for this incompatibility between LLVM and musl should be fixing the C++ compiler itself, but that’s another big story…

and [here](https://blogs.gentoo.org/gsoc2016-native-clang/2016/06/26/a-few-thoughts-on-libc-and-_gnu_source/):

> This week I was trying to make libc++ work without `_GNU_SOURCE` predefined, which causes me some trouble when compiling LLVM against musl. As mentioned in my last post, g++/clang++ unconditionally predefines _GNU_SOURCE for any C++ code, because libstdc++/libc++ simply won’t work without it. This is an old and well-known issue [[1]](http://web.mit.edu/darwin/src/modules/gcc3/libstdc++-v3/docs/html/faq/#3_5), but unfortunately has never been fixed. This week I boldly tried to fix it for libc++, and failed 🙁

and [here](https://wiki.gentoo.org/wiki/Musl_porting_notes):

> musl is very strict regarding standards-conformance compared to the widely used GNU C Library (glibc). This means that many of the GNU extensions, as well as much of the backwards compatibility provided by glibc is completely removed, and applications that use these will often fail to build. Here are some pointers on getting software to compile with musl.

In summary, combining musl and `_GNU_SOURCE` seems like a forbidden recipe for trouble. 😕 

We could carve out an exclusion to handle the musl case (despite [no feature test macro to detect musl directly](https://wiki.musl-libc.org/faq.html#Q:_Why_is_there_no_%3Ccode%3E__MUSL__%3C/code%3E_macro?), where the XSI-compliant variant is present despite `_GNU_SOURCE` being set, which is normally not the case.

**Comment by jmikola** [comment]
@kevinAlbs: Thanks for catching that. That very issue was discussed in https://stackoverflow.com/q/41953104/162228, which I linked from [CDRIVER-4679](https://jira.mongodb.org/browse/CDRIVER-4679) and soon after forgot. Unfortunately, the Stack Overflow question was using C++ and ultimately worked around this using operator overloading.

**Comment by kevinAlbs** [comment]
> We could carve out an exclusion to handle the musl case (despite [no feature test macro to detect musl directly](https://wiki.musl-libc.org/faq.html#Q:_Why_is_there_no_%3Ccode%3E__MUSL__%3C/code%3E_macro?)), where the XSI-compliant variant is present despite `_GNU_SOURCE` being set, which is normally not the case.

I vote for carving out an exclusion to handle musl. Given repeated requests for Alpine, I expect Alpine may eventually be a tested platform in the future (filed CDRIVER-4696 to consider testing with Alpine in Evergreen).

**Comment by eramongodb** [comment]
Given our findings above, which are now documented in the [CMakeLists.txt](https://github.com/mongodb/mongo-c-driver/pull/1350/files#diff-71a4c7cf57df62b3b3d8bb547bcf79c5fafb5accb4a6608d696ccdf43e17ad6aR122) file, gave up on using feature test macros and instead opted to use CMake's `CheckCSourceCompiles` module to obtain the definitive truth. Verified by [this patch](https://spruce.mongodb.com/version/64baf49bd1fe07e87a839b4f/).

File: src/libbson/CMakeLists.txt [hunk 2]
```diff
@@ -119,6 +119,38 @@ else ()
...
+      BSON_HAVE_XSI_STRERROR_R
+      FAIL_REGEX "int-conversion"
+   )
+endif ()
+
+if (BSON_HAVE_XSI_STRERROR_R)
+   set(BSON_HAVE_XSI_STRERROR_R 1)
+else()
+   set(BSON_HAVE_XSI_STRERROR_R 0)
+endif ()
```
Comment (jmikola) [inline src/libbson/CMakeLists.txt:152]: > gave up on using feature test macros and instead opted to use CMake's CheckCSourceCompiles module to obtain the definitive truth.

This does seem like the most reliable solution. Since this solution won't apply to the PHP driver, I created [PHPC-2263](https://jira.mongodb.org/browse/PHPC-2263) to port this to Autotools.

**Review by kevinAlbs — APPROVED** [review]
Confirmed latest changes build with Alpine in docker.
